File: 01-Programming-Languages/Go/Algorithms/Searching/binary_search.go
package main

import "fmt"

func binarySearch(arr []int, target int) int {
    left, right := 0, len(arr)-1

    for left <= right {
        mid := left + (right-left)/2

        if arr[mid] == target {
            return mid
        } else if arr[mid] < target {
            left = mid + 1
        } else {
            right = mid - 1
        }
    }

    return -1
}

func main() {
    arr := []int{2, 3, 4, 10, 40}
    target := 10
    result := binarySearch(arr, target)
    
    if result == -1 {
        fmt.Println("Element not found")
    } else {
        fmt.Printf("Element found at index %d\n", result)
    }
}

File: 01-Programming-Languages/Kotlin/Algorithms/Searching/binary_search.kt
fun binarySearch(arr: IntArray, target: Int): Int {
    var left = 0
    var right = arr.size - 1

    while (left <= right) {
        val mid = left + (right - left) / 2

        when {
            arr[mid] == target -> return mid
            arr[mid] < target -> left = mid + 1
            else -> right = mid - 1
        }
    }

    return -1
}

fun main() {
    val arr = intArrayOf(2, 3, 4, 10, 40)
    val target = 10
    val result = binarySearch(arr, target)

    if (result == -1) {
        println("Element not found")
    } else {
        println("Element found at index $result")
    }
}